### PR TITLE
Prototype peer network, binary protocol, block structure & store, etc.

### PR DESCRIPTION
**Pull request by zzim2x**
I've written some prototype codes.

This PR includes the following features.

* simple peer network
* binary protocol using msgpack (or something else)
* simple block strucuture & store
* catch up scenario
* etc.

### CODE REVIEW

File: pkg/rawdb/db.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+type Database interface {
+	Get(key []byte) ([]byte, error)
+
+	Has(key []byte) (bool, error)
+
+	Put(key []byte, value []byte) error
+
+	Delete(key []byte) error
+}
```
Comment (Geod24) [inline pkg/rawdb/db.go]: We also already have a LevelDB module, why do we need a new one ?

**Comment by zzim2x** [comment]
@Geod24 You're right. This PR's code is moved from personal space to this repo. So, there are same features.

I didn't spend a lot of time and made simple. If no one is interested, I'll close it. It's just prototype to check if you guys are interested.

File: Gopkg.lock [hunk 2]
```diff
@@ -34,6 +34,18 @@
...
+  name = "github.com/go-kit/kit"
+  packages = ["log"]
+  revision = "ca4112baa34cb55091301bdc13b1420a122b1b9e"
+  version = "v0.7.0"
+
+[[projects]]
+  name = "github.com/go-logfmt/logfmt"
+  packages = ["."]
+  revision = "390ab7935ee28ec6b286364bba9b4dd6410cb3d5"
+  version = "v0.3.0"
```
[the remark this answers is not in the document]
Comment (zzim2x) [reply Gopkg.lock]: I separated because it seemed to help reviewers understand changes. I'll squash them.

File: pkg/network/tcp/handshake.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	var err error
+	var their *message.HelloMessage
+	our := &message.HelloMessage{
+		ProtocolVersion: message.ProtocolVersion,
+		PubKey:          o.params.PubKey,
+		Timestamp:       uint32(time.Now().Second()),
+		Network:         o.params.NetworkPassphrase,
+	}
+
+	if inbound {
```
Comment (sconeman) [inline pkg/network/tcp/handshake.go]: What is the meaning of `inbound`?

**Comment by spikeekips** [comment]
@zzim2x I can understand what this PR trying to do, but it will be better that this PR needs more explanations in PR description.

File: pkg/network/tcp/handshake.go (hunk 3 above)
Comment (zzim2x) [reply pkg/network/tcp/handshake.go]: `inbound` is a incoming connection by accept loop.

File: pkg/wire/msgpack.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	t := reflect.TypeOf(m)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	if t.Kind() != reflect.Struct {
+		panic(errors.Errorf("%s type is not supported", t.Kind()))
+	}
```
Comment (sconeman) [inline pkg/wire/msgpack.go]: Is the part to confirm the message type?

File: Gopkg.lock (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmcm2222) [reply Gopkg.lock]: when I try to executing it, it gave me error. ```dep ensure```. 

```
Solving failure: No versions of boscoin.io/sebak met constraints:
	master: Could not introduce boscoin.io/sebak@master due to multiple problematic subpackages:
	Subpackage boscoin.io/sebak/pkg/core/dbi is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/network is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/rawdb is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/wire/message is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/wire is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/catchup is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/consensus is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/core is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/core/types is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/network/tcp is missing. (Package is required by (root).)	
Subpackage boscoin.io/sebak/pkg/support/logger is missing. (Package is required by (root).)
```

**Pull request closed without merging**